### PR TITLE
refactoring abstractions

### PR DESCRIPTION
**Pull request by Febius**

### CODE REVIEW

File: src/Parser/Element/ElementParserAbstract.php [hunk 1]
```diff
@@ -4,28 +4,20 @@
 
 use SurveyJsPhpSdk\Model\Element\ElementInterface;
 
-abstract class ElementParserAbstract
+abstract class ElementParserAbstract implements ElementParserInterface
 {
-    protected function configure(ElementInterface $element, \stdClass $data): ElementInterface
+
+    public function configure(ElementInterface $element, \stdClass $data): ElementInterface
```
Comment (astronati) [inline src/Parser/Element/ElementParserAbstract.php]: Public does not make sense. It's not called from the outside

File: src/Parser/Element/ElementParserAbstract.php [hunk 2]
```diff
@@ -4,28 +4,20 @@
...
-        }
-
-        if (isset($data->enableIf)) {
-            $element->setEnableIf($data->enableIf);
-        }
-
         return $element;
     }
 
-    abstract public function parse(ElementInterface $element, \stdClass $data): ElementInterface;
```
Comment (astronati) [inline src/Parser/Element/ElementParserAbstract.php:30]: Having an abstract here allows u to force to implement it for any extended class

File: src/Parser/SurveyResultParser.php [hunk 3]
```diff
@@ -50,20 +49,18 @@ public function parse(TemplateModel $survey, string $jsonResults): array
...
-//        }
-//
-//        return false;
+        /**
+         * @var PageModel $page
+         */
+        foreach($pages as $page){
+
+            foreach($page->getElements() as $element){
+                if($element instanceOf ElementAbstract && $element->isValidResult($result)) {
```
Comment (astronati) [inline src/Parser/SurveyResultParser.php]: you just need one element valid to return true. Are you sure?

File: src/Parser/Element/ElementParserInterface.php [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+
+namespace SurveyJsPhpSdk\Parser\Element;
+
+
+use SurveyJsPhpSdk\Model\Element\ElementInterface;
+
+interface ElementParserInterface
+{
+    public function configure(ElementInterface $element, \stdClass $data): ElementInterface;
```
Comment (astronati) [inline src/Parser/Element/ElementParserInterface.php]: I don't expect to have a configure as public.

File: src/Parser/Element/ChoiceElementParserAbstract.php [hunk 5]
```diff
@@ -6,7 +6,7 @@
 use SurveyJsPhpSdk\Model\Element\ChoiceElementAbstract;
 use SurveyJsPhpSdk\Model\Element\ElementInterface;
 
-abstract class ChoiceElementParserAbstract extends ElementParserAbstract
+abstract class ChoiceElementParserAbstract extends DefaultElementParser
```
Comment (astronati) [inline src/Parser/Element/ChoiceElementParserAbstract.php]: Let's define a better name for that.

File: src/Parser/Element/DefaultElementParser.php [hunk 6]
```diff
@@ -2,12 +2,28 @@
 
 namespace SurveyJsPhpSdk\Parser\Element;
 
+use SurveyJsPhpSdk\Model\Element\ElementAbstract;
 use SurveyJsPhpSdk\Model\Element\ElementInterface;
 
 class DefaultElementParser extends ElementParserAbstract
 {
-    public function parse(ElementInterface $element, \stdClass $data): ElementInterface
+    public function configure(ElementInterface $element, \stdClass $data): ElementInterface
```
Comment (astronati) [inline src/Parser/Element/DefaultElementParser.php]: this should be the protected

File: src/Parser/Element/ElementParserInterface.php [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+
+namespace SurveyJsPhpSdk\Parser\Element;
+
+
+use SurveyJsPhpSdk\Model\Element\ElementInterface;
+
+interface ElementParserInterface
+{
+    public function parse(ElementInterface $element, \stdClass $data): ElementInterface;
```
Comment (astronati) [inline src/Parser/Element/ElementParserInterface.php]: The interface is GOOD!

File: src/Parser/SurveyResultParser.php [hunk 8]
```diff
@@ -22,24 +21,24 @@ class SurveyResultParser
...
-//            $resultsModels[] = $resultModel;
-//        }
-//
-//        return $resultsModels;
+        $resultsModels = [];
+
+        $results = (array)json_decode($jsonResults);
+
+        foreach($results as $question => $result){
+            $resultModel = new ResultModel();
```
Comment (astronati) [inline src/Parser/SurveyResultParser.php]: Let's try to move this in a Factory.

File: src/Parser/SurveyResultParser.php (hunk 3 above)
Comment (astronati) [reply src/Parser/SurveyResultParser.php]: :up:

File: src/Parser/SurveyResultParser.php [hunk 9]
```diff
@@ -50,20 +49,18 @@ public function parse(TemplateModel $survey, string $jsonResults): array
...
+        foreach($pages as $page){
+
+            foreach($page->getElements() as $element){
+                if($element instanceOf ElementAbstract && $element->isValidResult($result)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
```
Comment (astronati) [inline src/Parser/SurveyResultParser.php]: Then try to run `composer fixcode`

File: src/Parser/SurveyTemplateParser.php [hunk 10]
```diff
@@ -45,12 +49,40 @@ public function parse(string $jsonTemplate): TemplateModel
...
+
+    /**
+     * @param \stdClass $element
+     *
+     * @throws ElementTypeNotFoundException
+     * @throws MissingElementConfigurationException
+     *
+     * @return ElementConfiguration|null
+     */
+    private function getConfigForElement(\stdClass $element): ?ElementConfiguration
```
Comment (astronati) [inline src/Parser/SurveyTemplateParser.php]: Let's discuss about this

File: src/Parser/SurveyTemplateParser.php [hunk 11]
```diff
@@ -45,12 +49,40 @@ public function parse(string $jsonTemplate): TemplateModel
             $pageModel = PageFactory::create($page);
 
             foreach ($page->elements as $element) {
-                $pageModel->addElement(ElementFactory::create($element, $this->customConfigurations[$element->type]));
+
+                $config = $this->getConfigForElement($element);
```
Comment (astronati) [inline src/Parser/SurveyTemplateParser.php]: You have this https://github.com/Febius/surveyjs-php-sdk/pull/5/files#diff-cbdd47fe1306b7884a36d6546ba42e9cR27,
why do you think is needed having this as well?

File: src/old/Parser/Element/CheckboxParser.php [hunk 12]
```diff
@@ -11,7 +11,7 @@ class CheckboxParser
     /**
      * @param \stdClass $element
      *
-     * @return CheckboxElement
+     * @return CheckboxElementParser
```
Comment (astronati) [inline src/old/Parser/Element/CheckboxParser.php]: you can delete old folder :D

File: src/Factory/ElementFactory.php [hunk 13]
```diff
@@ -30,13 +40,13 @@ public static function create(\stdClass $element, ?ElementConfiguration $configu
     {
         switch ($element->type) {
             case self::CHECKBOX_TYPE:
-                $parser = new DefaultChoiceElementParser();
+                $parser = new CheckboxElementParser();
```
Comment (astronati) [inline src/Factory/ElementFactory.php:47]: Why do u need a specific parser for all of these elements?

File: src/Factory/ElementFactory.php [hunk 14]
```diff
@@ -20,6 +23,13 @@ class ElementFactory
     const RADIO_GROUP_TYPE = 'radiogroup';
     const RATING_TYPE = 'rating';
 
+    const KNOWN_TYPES = [
```
Comment (astronati) [inline src/Factory/ElementFactory.php]: From what I saw let's discuss about this, I think it's not needed

File: src/Parser/Element/CheckboxElementParser.php [hunk 15]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+
+namespace SurveyJsPhpSdk\Parser\Element;
+
+
+class CheckboxElementParser extends ChoiceSurveyElementParserAbstract
```
Comment (astronati) [inline src/Parser/Element/CheckboxElementParser.php]: Let's avoid Survey in the name

File: src/Parser/Element/ChoiceSurveyElementParserAbstract.php [hunk 16]
```diff
@@ -22,14 +22,19 @@ public function parse(ElementInterface $element, \stdClass $data): ElementInterf
...
+            $choiceData = $this->formatChoiceObject($value);
 
             $element->addChoice(ChoiceFactory::create($choiceData));
         }
 
         return $element;
     }
 
-    abstract protected function getChoicesData(\stdClass $data): array;
+    protected function getChoicesData(\stdClass $data){
```
Comment (astronati) [inline src/Parser/Element/ChoiceSurveyElementParserAbstract.php]: mmm why did u change this? In this way who will extend this class, he won't be forced to implement this class

File: src/Parser/Element/ElementParserAbstract.php [hunk 17]
```diff
@@ -4,28 +4,20 @@
 
 use SurveyJsPhpSdk\Model\Element\ElementInterface;
 
-abstract class ElementParserAbstract
+abstract class ElementParserAbstract implements ElementParserInterface
```
Comment (astronati) [inline src/Parser/Element/ElementParserAbstract.php:8]: Implementing Interface is fine! Good catch!

File: src/Parser/Element/ElementParserAbstract.php [hunk 18]
```diff
@@ -4,28 +4,21 @@
...
 {
-    protected function configure(ElementInterface $element, \stdClass $data): ElementInterface
+    protected $element;
```
Comment (astronati) [inline src/Parser/Element/ElementParserAbstract.php:10]: Why an element should have a protected variable called `element`?

File: src/Factory/ElementFactory.php [hunk 19]
```diff
@@ -3,49 +3,68 @@
...
             case self::RATING_TYPE:
                 $parser = new RatingElementParser();
-                return $parser->parse(new RatingElement(), $element);
+                return $parser->parse($element);
             default:
-                if (null === $configuration) {
-                    throw new MissingElementConfigurationException($element->type);
+                if ($element->type === $configuration->getType()) {
+
+                    if(get_class($model = $configuration->getParser()->parse($element)) !== get_class($configuration->getElement())){
```
Comment (astronati) [inline src/Factory/ElementFactory.php]: Add a comment to explain what you are doing

File: src/Parser/Element/CheckboxElementParser.php (hunk 15 above)
Comment (astronati) [reply src/Parser/Element/CheckboxElementParser.php]: IMO the name of classes does not need the word "Survey". Everything is around it. You already have it in the namespace

File: src/Parser/Element/ElementParserAbstract.php [hunk 20]
```diff
@@ -4,28 +4,21 @@
...
+
+    protected function configure(\stdClass $data): void
     {
         if (isset($data->name)) {
```
Comment (astronati) [inline src/Parser/Element/ElementParserAbstract.php]: you expect to not save if name is not set. Is this right?
From what I see in `ElementInterface` you expect to always have name, so this means I would expect an error if this property is not set

File: src/Factory/ElementFactory.php [hunk 21]
```diff
@@ -3,49 +3,68 @@
...
 use SurveyJsPhpSdk\Model\Element\CommentElement;
 use SurveyJsPhpSdk\Model\Element\ElementInterface;
 use SurveyJsPhpSdk\Model\Element\RadioGroupElement;
 use SurveyJsPhpSdk\Model\Element\RatingElement;
+use SurveyJsPhpSdk\Parser\Element\CheckboxElementParser;
+use SurveyJsPhpSdk\Parser\Element\CommentElementParser;
 use SurveyJsPhpSdk\Parser\Element\DefaultChoiceElementParser;
-use SurveyJsPhpSdk\Parser\Element\DefaultElementParser;
+use SurveyJsPhpSdk\Parser\Element\DefaultSurveyElementParserAbstract;
+use SurveyJsPhpSdk\Parser\Element\RadiogroupElementParser;
```
Comment (astronati) [inline src/Factory/ElementFactory.php]: Let's call it Radio`G`roupElementParser

File: src/Parser/Element/CheckboxElementParser.php [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+use SurveyJsPhpSdk\Model\Element\CheckboxElement;
+use SurveyJsPhpSdk\Model\Element\ElementInterface;
+
+class CheckboxElementParser extends ChoiceElementParserAbstract
+{
+    public function parse(\stdClass $data): ElementInterface
+    {
+        $this->element = new CheckboxElement();
+
+        return parent::parse($data);
```
Comment (astronati) [inline src/Parser/Element/CheckboxElementParser.php]: Is the parent static?

File: src/Parser/Element/ChoiceElementParserAbstract.php [hunk 23]
```diff
@@ -5,31 +5,50 @@
...
-            // TODO Try to define a shared way with RatingElementParser::getChoicesData
-            $choiceData = ! is_object($value) ? (object)['text' => $value, 'value' => $value] : $value;
+            $choiceData = $this->formatChoiceObject($value);
 
-            $element->addChoice(ChoiceFactory::create($choiceData));
+            $this->element->addChoice($choiceParser->parse($choiceData));
         }
 
-        return $element;
+        return $this->element;
```
Comment (astronati) [inline src/Parser/Element/ChoiceElementParserAbstract.php:34]: Why a parser is not static? Why does it have a state?

Comment (Febius) [reply src/Parser/Element/ChoiceElementParserAbstract.php:34]: This way the dev implementing new survey elements or custom elements needs to instantiate the correct model inside the parser. This is necessary only for element parsers, not for all parsers

**Pull request merged**